### PR TITLE
Do not suggest using | in let assignments

### LINKED ISSUE #61573: Invalid suggestion to use a vertical bar in `let a, b = x;`
Introduced in #57008, the [following program](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=4043a87379488b48fbfa3a2bad796e88)
```
fn main() {
    let a, b = x;
}
```
yields the following message:
```
error: unexpected `,` in pattern
 --> src/main.rs:2:10
  |
2 |     let a, b = x;
  |          ^
help: try adding parentheses to match on a tuple..
  |
2 |     let (a, b) = x;
  |         ^^^^^^
help: ..or a vertical bar to match on multiple alternatives
  |
2 |     let a | b = x;
  |         ^^^^^
```

However, the suggestion `let a | b = x;` is nonsensical.

### PR DESCRIPTION
**Pull request by estebank**
Fix #61573.

r? @petrochenkov

### CODE REVIEW

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -3877,20 +3877,29 @@ impl<'a> Parser<'a> {
...
+                    }),
                     format!("({})", seq_snippet),
                     Applicability::MachineApplicable
-                ).span_suggestion(
-                    seq_span,
-                    "..or a vertical bar to match on multiple alternatives",
-                    format!("{}", seq_snippet.replace(",", " |")),
-                    Applicability::MachineApplicable
                 );
+                if is_match_arg {
```
Comment (Centril) [inline src/libsyntax/parse/parser.rs]: Please leave a `FIXME(#54883)` here and above for undoing this once we have or-patterns.
cc https://github.com/rust-lang/rust/issues/54883 cc @dlrobertson

**Review by Centril — CHANGES_REQUESTED** [review]

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (estebank) [reply src/libsyntax/parse/parser.rs]: My thinking is that this suggestion will _extremely rarely_ be correct in `let` statements. Anyone that want's to do `let x | y = foo();` will try _that_, whereas the typo `let x, y` is incredibly more likely to have meant a tuple as opposed to the (currently unsupported) `let x | y`.

Comment (Centril) [reply src/libsyntax/parse/parser.rs]: Hmm; I suppose you're right about that given that `a | b` won't type-check in any case.

**Comment by petrochenkov** [comment]
I'd prefer not to do this, but rather address the "`a | b` won't type-check in any case" using a simple heuristic.

If we have a comma separated list `PAT, PAT, PAT`, parse it until the end, then decide how to interpret it.
- Collect a set of lower-case identifiers for each `PAT` (with a small visitor implementing only `visit_ident`).
- If those sets are identical, recommend only `PAT | PAT | PAT`.
- If none of those sets intersect with each other, recommend only `(PAT, PAT, PAT)`.
- Otherwise recommend both and let the user disambiguate.

This work is done only in the error case, so it's not a performance issue.

**Comment by estebank** [comment]
@petrochenkov I like the idea, but it'll take me a while making the time to implement it.

**Pull request closed without merging**